src/main/scala/net/snowflake/spark/snowflake/io/CloudStorageOperations.scala

```diff
@@ -1860,3 +1880,184 @@ object StorageUtils {
   }
 }
 
+private[io] class S3UploadOutputStream(s3Client: AmazonS3Client,
```

Comment:
Quick question, if the job fails part way and some of the parts are uploaded already, is there any impact?

Reply:
Below description from AWS document addresses your question. The updated parts reside in the stage and AWS suggest to configure some strategy to clean up it. This is not a problem for our use case because the internal stage is temp stage, so it will be cleaned up in the end. 

=====================
After you initiate a multipart upload, you begin uploading parts. Amazon S3 stores these parts, but it creates the object from the parts only after you upload all of them and send a successful request to complete the multipart upload (you should verify that your request to complete multipart upload is successful). Upon receiving the complete multipart upload request, Amazon S3 assembles the parts and creates an object.

If you don't send the complete multipart upload request successfully, Amazon S3 will not assemble the parts and will not create any object. Therefore, the parts remain in Amazon S3 and you pay for the parts that are stored in Amazon S3. As a best practice, we recommend you configure a lifecycle rule (using the AbortIncompleteMultipartUpload action) to minimize your storage costs.

Reply:
https://docs.aws.amazon.com/AmazonS3/latest/dev/mpuoverview.html#mpu-stop-incomplete-mpu-lifecycle-config